ring/ownership_priority_queue.go

```diff
@@ -0,0 +1,139 @@
+}
+
+type ringToken struct {
+	token     uint32
+	prevToken uint32
+}
+
+func (rt ringToken) key() int {
+	return int(rt.token)
+}
+
+func (rt ringToken) String() string {
+	return fmt.Sprintf("[token: %d, prevToken: %d]", rt.token, rt.prevToken)
+}
+
+type ownershipInfo[T ringItem] struct {
+	item      T
+	ownership float64
+}
+
+func newRingTokenOwnershipInfo(token, prevToken uint32) ownershipInfo[ringToken] {
+	ownership := float64(tokenDistance(prevToken, token))
+	return ownershipInfo[ringToken]{
+		ownership: ownership,
+		item: ringToken{
+			token:     token,
+			prevToken: prevToken,
+		},
+	}
+}
+
+func newRingInstanceOwnershipInfo(instanceID int, ownership float64) ownershipInfo[ringInstance] {
+	return ownershipInfo[ringInstance]{
+		ownership: ownership,
+		item: ringInstance{
+			instanceID: instanceID,
+		},
+	}
+}
+
+// ownershipPriorityQueue is a max-heap, i.e., a priority queue
+// where items with a higher priority will be extracted first.
+// Namely, items with a higher ownership have a higher priority.
+// In order to guarantee that 2 instances of ownershipPriorityQueue
+// with the same items always assign equal priorities to equal items,
+// in the case of items with equal ownership, we rely on the
+// order of item ids.
+type ownershipPriorityQueue[T ringItem] struct {
+	items []ownershipInfo[T]
+}
+
+func newPriorityQueue[T ringItem](capacity int) ownershipPriorityQueue[T] {
+	return ownershipPriorityQueue[T]{
+		items: make([]ownershipInfo[T], 0, capacity),
+	}
+}
+
+func (pq *ownershipPriorityQueue[T]) Len() int {
+	return len(pq.items)
+}
+
+func (pq *ownershipPriorityQueue[T]) Swap(i, j int) {
+	pq.items[i], pq.items[j] = pq.items[j], pq.items[i]
+}
+
+func (pq *ownershipPriorityQueue[T]) Less(i, j int) bool {
+	if pq.items[i].ownership == pq.items[j].ownership {
+		// In order to guarantee the stability, i.e., that the same instanceID and zone as input
+		// always generate the same slice of tokens as output, we enforce that by equal ownership
+		// higher priority is determined by the order of ids.
+		return pq.items[i].item.key() > pq.items[j].item.key()
+	}
+	// We are implementing a max-heap, so we are using > here.
+	// Since we compare float64, NaN values must be placed at the end.
+	return pq.items[i].ownership > pq.items[j].ownership || (math.IsNaN(pq.items[j].ownership) && !math.IsNaN(pq.items[i].ownership))
+}
+
+// Push implements heap.Push(any). It pushes the element item onto ownershipPriorityQueue.
+// Time complexity is O(log n), where n = Len().
+func (pq *ownershipPriorityQueue[T]) Push(item any) {
+	ownershipInfo := item.(ownershipInfo[T])
+	pq.items = append(pq.items, ownershipInfo)
+}
+
+// Pop implements heap.Pop(). It removes and returns the element with the highest priority from ownershipPriorityQueue.
+// Time complexity is O(log n), where n = Len().
+func (pq *ownershipPriorityQueue[T]) Pop() any {
+	old := pq.items
+	n := len(old)
+	item := old[n-1]
```

Comment:
I'm missing something. The heap keeps the minimum element in the tree is the root, so at index 0. Since Less() is implemented in "inverse order" ("less" returns true for the item with the highest priority), at index 0 we expect the item with the highest priority. Why do we look at the last one instead?

Reply:
Here is the [example](https://pkg.go.dev/container/heap#:~:text=Example,-(PriorityQueue)%20%C2%B6) provided by container/heap that I used as a model. Basically every custom implementation is used [at the end](https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/container/heap/heap.go;l=63) of [heap.Pop](https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/container/heap/heap.go;l=59) implementation that I call when I want to remove the element with the highest ownership. At line [61](https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/container/heap/heap.go;l=61) the highest element and the element at index len()-1 are swapped. So it means that the highest element is now at index len()-1 . Then at line 62, the down method is called, which “heapifies” elements [0-len()-1) (I used “)” to show that the last one is excluded). And that’s why the custom Pop implementation needs to return element len()-1. Basically, this doesn’t change either when you implement min or max heap. The only difference should be the Less method, which in the case of max heap should invert the order.